api/v1alpha1/proxmoxmachine_types.go

```diff
@@ -203,6 +203,110 @@ type NetworkSpec struct {
+	IPv6PoolRef *corev1.TypedLocalObjectReference `json:"ipv6PoolRef,omitempty"`
+
+	// DNSServers contains information about nameservers to be used for this interface.
+	// If this field is not set, it will use the default dns servers from the ProxmoxCluster.
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	DNSServers []string `json:"dnsServers,omitempty"`
+}
+
+// RouteSpec describes an IPv4/IPv6 Route.
+type RouteSpec struct {
+	// To is the subnet to be routed.
+	// +optional
+	To string `json:"to,omitempty"`
+	// Via is the gateway to the subnet.
+	// +optional
+	Via string `json:"via,omitempty"`
+	// Metric is the priority of the route in the routing table.
+	// +optional
+	Metric uint32 `json:"metric,omitempty"`
+	// Table is the routing table used for this route.
+	// +optional
+	Table uint32 `json:"table,omitempty"`
+}
+
+// RoutingPolicySpec is a linux FIB rule.
+type RoutingPolicySpec struct {
+	// To is the subnet of the target.
+	// +optional
+	To string `json:"to,omitempty"`
+
+	// From is the subnet of the source.
+	// +optional
+	From string `json:"from,omitempty"`
+
+	// Table is the routing table id.
+	// +optional
+	Table uint32 `json:"table,omitempty"`
+
+	// Priority is the position in the ip rule fib table.
+	// +kubebuilder:validation:Maximum=4294967295
+	// +kubebuilder:validation:XValidation:message="Cowardly refusing to insert fib rule matching kernel rules",rule="(self > 0 && self < 32765) || (self > 32766)"
+	// +optional
+	Priority uint32 `json:"priority,omitempty"`
+}
+
+// VrfDevice defines Virtual Routing Flow devices.
+type VrfDevice struct {
+	// Interfaces is the list of proxmox network devices managed by this virtual device.
+	Interfaces []string `json:"interfaces,omitempty"`
+
+	// Name is the virtual network device name.
+	// must be unique within the virtual machine.
+	// +kubebuilder:validation:MinLength=3
+	Name string `json:"name"`
+
+	// Table is the ID of the routing table used for the l3mdev vrf device.
+	// +kubebuilder:validation:Maximum=4294967295
+	// +kubebuilder:validation:XValidation:message="Cowardly refusing to insert l3mdev rules into kernel tables",rule="(self > 0 && self < 254) || (self > 255)"
+	Table uint32 `json:"table"`
+
+	// InterfaceConfig contains all configurables a network interface can have.
+	// +optional
+	InterfaceConfig `json:",inline"`
+
+	// Routes are the routes associated with the l3mdev policy.
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	Routes []RouteSpec `json:"routes,omitempty"`
+
+	// RoutingPolicy is the l3mdev policy inserted into FiB.
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	RoutingPolicy []RoutingPolicySpec `json:"routing-policy,omitempty"`
+}
+
+// VirtualNetworkDevices defines linux software networking devices.
+type VirtualNetworkDevices struct {
+	// Definition of a Vrf Device.
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	VRFs []VrfDevice `json:"vrfs,omitempty"`
```

Comment:
Why is this not part of additional devices?

Reply:
Can those VRFs be added with command-line?

Reply:
That's because the rules for validation would need reworking. Additional devices are strictly "physical" at the moment.
It makes sense to split the setup into physical and software defined, as this is effectively how netplan handles it too.

Can VRFs be added by CLI? Yes. Unfortunately, we do not have the information to do so. This requires knowing the interface name. We do not know the actual interface name until after the machine boots, as this is not deterministic (interface names are made from rules matching pci-e addresses). Netplan arranges those things for us via mac address matching.

The short gist of it is: we're using netplan and we're not reinventing the wheel.